README.md
## NAME

    BoundaryFinder - Exon/Intron Boundary Finder Module

## SYNOPSIS

      use BoundaryFinder;

## DESCRIPTION

    The BoundaryFinder uses consensus sequences to find and score
    boundaries in a sequence. This was developed for finding exon
    boundaries, but it can be used with any consensus sequence.

    The boundary finder works best when you finding boundaries in 
    small sequences. This is because the number of matches is often
    large, especially if there are no "100%" positions. (e.g. 100% 
    of nucleotides have to be an "A" at position x, or it does not 
    match)

## OVERVIEW

    The general procedure of using the library to find exon boundaries
    is as follows:

### 1. Use the BlastManager to blast your query sequence agaist the
          subject database.

          This will give a BlastResults oject, or a set of blast hits.

### 2. Determine what BlastResults or hits you want use.

          Sometimes blast returns hits that are not relevant to your
          query. These could be weak matches that blast returned that
          do not make sense with your query. You can trim these out 
          by changing the blast parameters (such as a minimum "e" value)
          or iterating through the BlastResults and removing the 
          unwanted hits.

### 3. Use BoundaryFinder to find the boundaries of the blast hits

          Each query is a pair of sequences representing the 5` and 3`
          ends of a blast hit.

          This can be automatically done by passing a BoundaryResults 
          object, or done manually. It will create a query with a 50 
          extra nucleotides from each end of the edges of the blast hit. 
          You can easily specify a buffer size other than 50.

          ***************************************************************
          E.g. 

          This hastily done ascii example shows how the BoundaryFinder will
          create queries automatically from a hit. Given the blast result below,
          a sequence is created by using 50 nucleotides upstream (5') to the end
          of the blast hit (3' end). This is also done for the 3' end, with 
	  50 NT after the end of the hit.

          Query   ATGATGATGATGATG
                  | |||||||||||    ... and so on
          Sub     ACGATGATGATGACC
                 
          |< -50  ^  ------------- To the end of the blast hit --------------->|

          ^ = Points to the 5' edge of the blast result.

          ***************************************************************

          This can be done manually too, and it takes only a little more 
          work. You will just have to figure out the buffers for each hit,
          and add them with the BoundaryFinder::addQuery method.

### 4. Run the BoundaryFinder to get a BoundaryResults object.

          The BoundaryResults object contains all the non-zero scores, and
          their positions in the blast database for easy retrieval. It is 
          simply a wrapper of two hashes that contain the position as keys,
          and score as the value.

## USAGE

     Coming Soon...

## AUTHOR

        Nathan Elmore
        nate@elmoren.com
        http://www.elmoren.com

## COPYRIGHT

	

## SEE ALSO
   
	Blast+ 2.2.28

=cut


Changes
Revision history for Perl module BoundaryFinder

0.01 Sun Jan 20 21:27:02 2013
    - original version; created by ExtUtils::ModuleMaker 0.51




lib/BlastManager.pm
package BlastManager;

use warnings;
use strict;

# For determining blast+ tools availability
use Carp;
use File::Which;

#################### main pod documentation begin ###################

=head1 NAME

BlastManager - Blast Tools Wrapper

=head1 SYNOPSIS

  use BlastManager;

  my $BM = BlastManager->new(-blast-db => "path/to/blast/db");

  # FASTA formatted query file
  my $query_path = "path/to/query.fna";

  my $blast_results = $BM->blastn(-query => $query_path);
  ... or
  my $blast_results = $BM->tblastn(-query => $query_path);
  ... or
  my $blast_results = $BM->blastx(-query => $query_path);

=head1 DESCRIPTION

The BlastManager is a wrapper class for the blast command line tools. You specify the source
database, the blast type (blastn, tblastn, etc) and the query. It then gives you a BlastResults
object from the result. 

The BlastResults object is a wrapper to the outputs of a blast query, providing methods to get
sequences of the blast, the score, and the human readable output. See BlastResults.

The BlastManager was made specifically for the BoundaryFinder. It can be used for other purposes,
but it is not as flexible as teh command line tools.

=cut

#################### main pod documentation end ###################

=head1 METHODS

=cut

#################### subroutine header begin #######################

=head2 new

 Usage     : my $BM = BlastManager->new();

 Purpose   : Creates new BlastManager

 Returns   : A BlastManager

 Argument  : 

 Comment   : This class is a wrapper class for the blast command
             line interface.

=cut

sub new
{
    my $class = shift;
    
    my $self = {
	-blast_db => "",
	-evalue => 1,
	@_
    };

    bless ($self, ref ($class) || $class);

    # Test if Blast Exists on the system.
    if ( which('blastn') ) {
	print "Warning: Blast+ tools not found in PATH\n";
    }

    return $self;
}

=head2 blast_db

 Usage     : my $blast_db = $BM->blast_db(-db => "some/path/blastdb_name");
           : my $blast_db = $BM->blast_db();

 Purpose   : Getter/Setter for the current blastdb parms

 Returns   : The blastdb path, 

 Argument  : -db: database path and name as string

 Comment   : 

=cut

sub blast_db
{
    my $self = shift;
    
    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;
    
    $self->{-blast_db} = $params{-db} if defined ($params{-db});
    return $self->{-blast_db}; 
}

=head2 blastn

 Usage     : my $blast_results = $BM->blastn({-query => "queryPath"});

 Purpose   : Runs a blastn query with the given query path or sequence

 Returns   : A BlastResults object

 Argument  : -query: Path to a blast query Database
             -query-seq: a seqence of NT

 Comment   : If both arguments are provided, it uses the query path method and gives a warning

=cut

sub blastn
{
    
}

=head2 blastx

 Usage     : my $blast_results = $BM->blastx(-query => "queryPath");

 Purpose   : Runs a blastx query with the given query path or sequence

 Returns   : A BlastResults object

 Argument  : -query: Path to a blast query Database
             -query-seq: a seqence of NT

 Comment   : If both arguments are provided, it uses the query path method and gives a warning

=cut

sub blastx
{
    croak "BlastManager Error: Not supported yet";
}

=head2 evalue

 Usage     : my $evalue = $BM->evalue(-e => 10e-1);
             my $evalue = $BM->evalue;

 Purpose   : Gets and sets the evalue for the queries

 Returns   : A BlastResults object

 Argument  : -query: Path to a blast query Database
             -query-seq: a seqence of NT

 Comment   : The expected value for saving hits (evalue) is useful when want to remove
             hits that do not have enough matches. Or, it can be used to try and acquire
             more hits that have less matches. 

             See http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect

=cut

sub evalue
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-evalue} = $params{-e} if defined ($params{-e});
    return $self->{-evalue};
}


=head2 get_sequence

 Usage     : my $sequence = $BM->get_sequence(-db => "dmel", -seq_id => "seq_id", -start => 1, -end => 10);

 Purpose   : gets a sequence from the database

 Returns   : A fasta sequence from the blast database

 Argument  : [-db] : The database to get it from (not required if -blast_db set)
             -seq_id : The sequence id of the database
             -start : The start index of the sequence
             -end : then end index

 Comment   :


=cut

sub get_sequence
{
    my $self = shift;

    croak "BlastManager->get_sequence(): Illegal parameter list has odd number of values" if @_ % 2;
    
    my %args = @_;

    for my $required (qw{ -seq_id -start -end }) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $args{$required}; 
    }
    
    my ($seq, $db, $range, $seq_id);
    
    if (exists $args{-db}) {
	$db = $args{-db};
    } 
    elsif (exists $self->{-blast_db} && $self->{-blast_db} ne "")
    {
	$db = $self->{-blast_db};
    }
    else
    {
	croak "BlastManager->get_sequence(): No blast db to query";
    }

    # Regular order
    $seq_id = $args{-seq_id};
    if ($args{-start} <= $args{-end}) {
	
	$range = "$args{-start}-$args{-end}";
	$seq = `blastdbcmd -db $db -range $range -entry $seq_id`;

	my @lines = split /\n/, $seq;
        shift @lines;
	$seq = join "", @lines;
    }
    # Reversed Complement Order
    else {

	$range = "$args{-end}-$args{-start}";
	$seq = `blastdbcmd -db $db -range $range -entry $args{-seq_id}`;
	
	my @lines = split /\n/, $seq;
        shift @lines;
	$seq = join "", @lines;

	$seq = reverse($seq);    
	# complement the reversed DNA sequence
	$seq =~ tr/ABCDGHMNRSTUVWXYabcdghmnrstuvwxy/TVGHCDKNYSAABWXRtvghcdknysaabwxr/;
    }

    return $seq;
}

=head2 tblastn

 Usage     : my $blast_results = $BM->tblastn(-query => "queryPath");

 Purpose   : Runs a tblastn query with the given path to query

 Returns   : A BlastResults object, or croaks if error in query

 Argument  : -query: Path to a blast query file
             [-db] : Path to subject database. (Required if -blast_db not set)

 Comment   : Other args to come, maybe. Custom formats not supported, yet.

=cut

sub tblastn
{
    my $self = shift;

    croak "BlastManager->tblastn(): Illegal parameter list has odd number of values" if @_ % 2;
    
    my $query = "";
    my $db = "";
    my $parms .= "-evalue " . $self->evalue;
    my $fmt = ""; # Not currently supported
    my %args =	@_;
 
    for my $required (qw{ -query }) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $args{$required}; 
    }

    $query = $args{-query};

    if (exists $args{-db}) {
	$db .= $args{-db};
    }
    elsif ($self->{-blast_db} ne "") {
	$db .= $self->{-blast_db};
    }
    else {
	croak "BlastManager->tblastn(): Error: No subject database";
    }

    my $csv   = `tblastn -query $query -db $db $parms -outfmt "10 $fmt"`;
    my $blast = `tblastn -query $query -db $db $parms`;

    # Check results, and if okay, Create BlastResults mand return it
    
    if ($csv =~ m/error/) {
	croak "BlastManager->tblastn(): Query error: \n $csv";
    }
    
    # Get the DB Source...?
    # my $source;

    # @TODO: Figure out how tot get the db source.
    # Currently assumes/requires the source to be $db.fas

    return BlastResults->new( -raw_csv => $csv, 
			      -raw_blast => $blast, 
			      -blast_type => 'tblastn',
			      -db_source => $db . '.fas',
			      -db_name => $db
	);

}

#################### subroutine header end ########################

#################### footer pod doc ###############################

=head1 BUGS

1. Database source file naming

Assumes the database fasta source file is the same as the database name with
a .fas extension. Not true because the user can name the database whatever
they want with makeblastdb. Looking for solution. It may have to force naming
convention.

=head1 AUTHOR

    Nathan Elmore
    nate@elmoren.com
    http://www.elmoren.com

=head1 COPYRIGHT

=cut

1;


lib/BlastResults.pm
package BlastResults;

use warnings;
use strict;

use Carp;
use File::Basename;
use Data::Dumper;
#################### main pod documentation begin ###################

=head1 NAME

BlastResults

=head1 SYNOPSIS

my $results = BlastResults->new();
$results->load_file(-file => "results.csv");

for (my $i = 0; $i < $results->size; $i++) {

    my $hit = get(-idx => $i);
    print $hit->{bitscore};

}

=head1 DESCRIPTION

The BlastResults object is a wrapper to the outputs of a blast query, providing 
methods to get data from the blast, the score, and the human readable output.
It also has methods to save and load previous blast results.

The BlastResults object contains a list of blast hits returned from the blast 
query. Iterate through the the results with and index and size.

The user may remove hits and save the results if they wish, but need to be careful if
they edit the data in the hits.

=head2 SAVE/LOAD RESULTS

The BlastResults object provides methods for saving and loading the results with
save_results and load_results. The user can remove unwanted hits before saving 
with the remove method. When the user calls "save". The save method expects a 
name and a directory.

 The BlastResults writes:

"<directory>/<name>.csv." - CSV file containing the output and other meta data

=cut

#################### main pod documentation end ###################

=head1 METHODS

=cut

#################### subroutine header begin #######################

#
# Defualt blast format used for parsing csv files when formats not specified.
#
my $BLAST_FMT =  [ 'qseqid', 'sseqid', 'pident', 'length', 'mismatch',
          	   'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 
                   'bitscore' ];



=head2 db_source

 Usage     : my $src = $BR->db_source();
             my $src = $BM->db_source(-file => "path/to/fasta");

 Purpose   : Gets and sets the blast source file. This is the path the FASTA 
             file that is used to create the subject database. 
             
             Setting this is not recommended unless you are creating a 
             BlastResults object without using a BlastManager instances.
             
 Returns   : Path to a Blast database source.

 Argument  : [-file] : Path to set

 Comment   : This is stored so that the Boundary Finder (or user) can extract a
             sequence from the database. It is a path to a FASTA file

=cut

sub db_source
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-db_source} = $params{-file} if defined ( $params{-file} );
    return $self->{-db_source};
}

=head2 blast_type

 Usage     : my $source = $BR->blast_type();
             my $evalue = $BR->blast_type(-type => "blastn");
 Purpose   : Gets and sets the blast type
             
 Returns   : Path to a Blast database source.

 Argument  : [-type] : Path to set

 Comment   : Settign this field not recommended if you use BlastManager to get
             the BlastResults object.

=cut

sub blast_type
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-blast_type} = $params{-type} if defined ( $params{-type} );
    return $self->{-blast_type};
}

=head2 csv_file

 Usage     : my $file = $BR->csv_file();
             my $file = $BR->csv_file(-file => "some/path/csv");

 Purpose   : Gets and sets the blast type             

 Returns   : Path to the a BlastResults CSV

 Argument  : [-file] : Path to set

 Comment   : Settign this field not recommended if you use BlastManager to get
             the BlastResults object.

=cut

sub csv_file
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-csv_file} = $params{-file} if defined ($params{-file});
    return $self->{-csv_file};
}

=head2 db_name

 Usage     : my $source = $BR->db_name();
             my $evalue = $BR->db_name(-name => "dmel");

 Purpose   : Gets and sets the blast database name

 Returns   : Path to a Blast database name

 Argument  : [-type] : Type to set

 Comment   : This is the name of the blast database. Which is not necessarily 
             the same as the database source name. You can have a fasta file,
             dmel.fas and name the database 'drosophila'. You can tell the
             name of the blase database by the name of the .nsq, .nhr, and .nin

=cut

sub db_name
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-db_name} = $params{-name} if defined ($params{-name});
    return $self->{-db_name};
}

=head2 get

 Usage     : my $hit = get(-idx => $i);

 Purpose   : Gets a hit at -idx

 Returns   : Hash containing the blast hit data

 Argument  : -idx : index of hit

 Comment   : 

=cut

sub get
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    croak "Required parameter '-idx' not passed to score_seq"
	if not exists $params{-idx};

    if ($params{-idx} < 0 || $params{-idx} >= $self->size) {
	croak "BlastResults->get(): Parameter [-idx] is out of bounds";
    }
    
    return $self->{_hits}->[$params{-idx}];
}


=head2 new

 Usage     : my $br = BlastManager::BlastResults->new(-csv_file => "path/to/results/csv");

 Purpose   : Creates a new blast results wrapper to access the data.

 Returns   : A BlastManager::BlastResults Object

 Argument  : [-raw_csv] : The output of a blast with '-outfmt 10'
           : [-format] : Reference to and array of outfmt csv format. Or it uses 
                         default
           : [-raw_blast] : The raw blast output (Default -outfmt). Available 
                            when using BlastManager
           : [-blast_type] : The blast type type used (blastx, tblastn, etc)
           : [-db_source] : Path to fasta file. Cannot extract sequences if this
                            is not set.
           : [-db_name] : Name of the blast database

 Comment   : There are multiple ways to instantiate this object:
           1. Use the blast manager to run a blast, and it will return a BlastResults 
              object
           2. Create a new BlastResults object and pass it raw csv data by adding 
              -outfmt 10 to your query. Use -raw_csv parameter or loat_text()
           3. Create a new BlastResults object and calling the load_file function.

=cut

sub new 
{
    my $class = shift;

    croak "BlastResults->new(): Illegal parameter list has odd number of values" if @_ % 2;

    my %params = @_;
    
    # default params
    my $self = {
	-raw_csv => "",
	-csv_file => "",
	-format => undef,
	-raw_blast => "",
	-blast_type => "",
	-db_source => "",
	-db_name  => ""
    };
    
    # Validate params by checking if they exist is self
    for my $attr (keys %params) {	
        croak "Invalid parameter '$attr' passed to '$class' constructor"
            unless defined ($self->{$attr});

	$self->{$attr} = $params{$attr};
    }


    # Init attributes user is not allowed to specify
    $self->{_size} = 0;
    $self->{_hits} = ();
    
    #
    # Init object if raw csv, or csv_file provided. CANNOT DO BOTH
    #    
    if ( $self->{-raw_csv} ne "" && $self->{-csv_file} ne "" ) {
	croak "BlastResults: does not support initialization of -raw_csv and -csv_file simultaneously" 
    }

    bless ($self, ref ($class) || $class);
    #
    # Init the blast data for the object
    #
    if ( $self->{-raw_csv} ne "" ) {
	$self->load_text( -text => $self->{-raw_csv} );
    }
    elsif ( $self->{-csv_file} ne "" ) {
	$self->load_file( -file => $self->{-csv_file} );
    }
    else {
	
    }


    return $self;
}


=head2 load_text

 Usage     : $blast_results->load_text(-text => $str);

 Purpose   : Loads resutls from raw csv text

 Returns   : Nothing

 Argument  : -text : CSV String
             [-format] : format (if not defualt)

 Comment   : 

=cut

sub load_text
{
    my $self = shift;
    
    croak "BlastResults->load_text(): Illegal parameter list has odd number of values" if @_ % 2;
    
    my $args = {
	-format => undef,
	@_
    };

    for my $required (qw{ -text }) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $args->{$required}; 
    }
    
    #
    # First clear out old data.
    #
    $self->{_size} = 0;
    $self->{_hits} = ();
    $self->{_file} = 0; # Not set by default
    
    if ( $args->{-format}) {
	$self->{-format} = $args->{-format};
    } 
    else {
	$self->{-format} = $BLAST_FMT;
    }

    #
    # Parse The CSV Data
    #

    # carp Dumper($self->{-format});

    foreach my $line (split (/\n/, $args->{-text})) {
	chomp $line;
	my $i = 0;
	my %hit = map { $self->{-format}->[$i++] => $_ } split (",", $line);
	
	push @{ $self->{_hits} }, \%hit;
	$self->{_size}++;
    }
    
    # carp Dumper($self->{_hits});

}

=head2 load_file

 Usage     : $blast_results->load_file(-file => $file);

 Purpose   : Loads blast results from a file

 Returns   : Nothing

 Argument  : [-file] : The path to csv file (If its not specified, will try
                       $self->{-csv_file} before giving an angry message

 Comment   : Checks the first line of the file for the format, if not present,
             uses the defualt of blasts 'outfmt 10' option

=cut

sub load_file
{
    my $self = shift;
    
    croak "BlastResults->load_file(): Illegal parameter list has odd number of values" if @_ % 2;
    
    my $args = {
	@_
    };
    
    # Check Optional File argument
    if (exists $args->{-file}) {
	$self->csv_file( -file => $args->{-file} );
    } 

    #
    # Look for .csv (required)
    #

    open CSV, "<", $self->csv_file or 
	croak "BlastResults::load_results Error: File not found: $!";

    #
    # Check for headers
    #
    
    my $line;
    my $csv_data = "";
    my $format;

    foreach my $line (<CSV>) {
	chomp $line;
	if ($line =~ m/DB_SOURCE=/) {
	    $self->{-db_source} = (split "=", $line)[1];
	}
	elsif ($line =~ m/DB_NAME=/) {
	    $self->{-db_name} = (split "=", $line)[1];
	}
	elsif ($line =~ m/FORMAT=/) {
	    $format = [ split /,/, $line ];
	}
	elsif ($line =~ m/BLAST_TYPE=/) {
	    $self->{-blast_type} = (split "=", $line)[1];
	}
	else {
	    $csv_data .= $line . "\n";
	}
    }


    close CSV;

    # Use defualt format if one was not specified in the file.
    $format = $BLAST_FMT if not $format;

    $self->load_text(-text => $csv_data, -format => $format);
}

=head2 raw_blast

 Usage     : my $file = $BR->raw_blast;
             my $file = $BR->raw_blast(-data => $data);

 Purpose   : Gets and sets the raw blast output

 Returns   : Raw Blast as string

 Argument  : [-data] : Raw blast string

 Comment   : 

=cut

sub raw_blast
{
    my $self = shift;

    croak "Illegal parameter list has odd number of values" if @_ % 2;
    my %params = @_;

    $self->{-raw_blast} = $params{-data} if defined ($params{-data});
    return $self->{-raw_blast};
}

=head2 save

 Usage     : $blast_results->save(-file => $file);

 Purpose   : Saves the results for future reference.

 Returns   : Nothing

 Argument  : -file : Desired file to write to. If you want to rewrite the file,
                     use $self->csv_file. Clobbers the old file, if it exists.

 Comment   : 

=cut

sub save
{

    my $self = shift;
    
    croak "BlastResults->save(): Illegal parameter list has odd number of values" if @_ % 2;
    
    my $args = {
	@_
    };

    # See if there is a file to save to
    $self->{_file} = $args->{-file} if exists $args->{-file};
    
    # open the file or croak
    if (exists $self->{_file}) {
	open OUT, ">$self->{_file}" or croak "BlastResults->save_results(): $!";
    }
    else {
	croak "BlastResults->save_results(): no file to write to.\n";
    }

    #
    # Write the headers
    #
 
    print OUT "DB_SOURCE=" . $self->db_source . "\n";
    print OUT "DB_NAME=" . $self->db_name . "\n";
    print OUT "BLAST_TYPE=" . $self->blast_type . "\n";
    
    my $fmt_string = join(",", @{ $self->{-format} }) if exists $self->{-format};
    print OUT "FORMAT=" . $fmt_string . "\n";
    
    #
    # Write the data
    #
    for (my $i = 0; $i < $self->size; $i++) {
	my @hit;
	foreach my $key ( @{ $self->{-format} } ) {
	    push @hit, $self->{_hits}->[$i]->{$key};
	}
	print OUT join(",", @hit) . "\n";
    }
    close OUT;
}

=head2 size

 Usage     : $num_result = $blast_results->size

 Purpose   : Gets number of blast hits

 Returns   : The total number of blast hits contained in the BlastResults

 Argument  : None

 Comment   : 

=cut

sub size
{
    my $self = shift;
    return $self->{_size};
}

#################### subroutine header end ########################

#################### footer pod doc ###############################

=head1 BUGS


=head1 AUTHOR

    Nathan Elmore
    nate@elmoren.com
    http://www.elmoren.com

=head1 COPYRIGHT

=head1 SEE ALSO

Blast Command Line Tools, BlastResults

=cut

1;


lib/BoundaryFinder.pm
package BoundaryFinder;

use warnings;
use strict;
use Carp;

BEGIN {
     require Exporter;
     use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
     our $VERSION     = '0.01';
     our @ISA         = qw(Exporter);

     our %EXPORT_TAGS = (
	 'all' => [ qw (
                         score_seq
                         score_hit
                  ) ]
     );
     our @EXPORT_OK   = qw( @{ $EXPORT_TAGS{'all'} } );
     our @EXPORT      = qw( score_seq );
}

#################### main pod documentation begin ###################

=head1 NAME

BoundaryFinder - Exon/Intron Boundary Finder

=head1 SYNOPSIS

  use BoundaryFinder;

  Coming soon...

=head1 DESCRIPTION

The boundary finder uses consensus sequences to find and score boundaries in a 
sequence. This was developed for finding exon boundaries, but it can be used 
with any consensus sequence.


=cut

#################### main pod documentation end ###################

=head1 METHODS

=cut

#################### subroutine header begin #######################

=head2 score_seq

 Usage     : my $scoresRef = BoundaryFinder::score_seq( 
                                               -seq => $seq, 
                                               -cons_seq => $cons_seq
                                              );

 Purpose   : Attempts to score a sequence hit
             BoundaryFinder::ConsensusSeq object.

 Argument  : -seq : A sequence to score
             -cons_seq : initialized BoundaryFinder::ConsensusSeq object
             [-min_score] : The minimum score to save (Defualt: 0)

 Returns   : A hash reference in which the key is the index of the start of
             the match, and the value is the score.

             $scores = {
               17 => 1.23e-08,
               50 => 7.23e-04,
               ...
             }

 Comment   : 

=cut

sub score_seq
{

    croak "Illegal parameter list has odd number of values" if @_ % 2;

    my %params = (
	-min_score => 0,
	@_
    );

    my %results;

    # Check required parameters
    for my $required (qw{ -seq -cons_seq }) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $params{$required};
    }
    
    # Use the cons seq to score small chunks and keep track of the hits above 
    # the min_score
    my $cons = $params{-cons_seq};
    for (my $i = 0; $i < length($params{-seq}); $i++) {
	my $sub_seq = substr($params{-seq}, $i, $cons->length);
	my $score = $cons->score_seq(seq => $sub_seq);

	# Add if it meets the minimum score
	if ($score > $params{-min_score}) {
	    $results{$i} = $score;
	}
    }

    return \%results;
}

=head2 score_hit

 Usage     : my $ref_to_hit = $blastResultsObj->get(-idx => 0);
             my $scoresRef = BoundaryFinder::score_hit ( 
                                               -hit => $ref_to_hit, 
                                               -db => $db_name,
                                               -cons_seq_5 => $cons_seq_5_prime,
                                               -cons_seq_3 => $cons_seq_3_prime,
                                               -padding => $padding_len,
                                               -min_sore => $min_score
                                              );

 Purpose   : Searches for Boundaries given a BlastResults hit. It attempts to 
             pull the sequence from the database to score it. 

             It uses two consensus sequences, one for the 5' and 3' end. The 
             padding value is optionally set to adjust the 'n' NTs upstream and 
             downstream of 'sstart' and 'send'. Otherwise it searches the whole 
             blast hit, with a default 50 NT upstream and downstream.

             [sstart - 50] to [send + 50]

             The boundary finder will ignore scores less than the optional 
             '-min_score' argument.

 Argument  : -hit
             -db : the database name
              -cons_seq_5 : The 5' consensus sequence (ConsensusSeq object)
             -cons_seq_3 : The 3' consensus sequence (ConsensusSeq object)
             [-padding] : The number of nucleotides to pad the hit on both the
                          5' and 3' ends of the blast hit. Otherwise it defaults
                          to 50 nucleotides.
             [-min_score] : Minimum score to keep in the results

 Returns   : A hash reference in which the key is the index of the start of
             the match, and the value is the score.

             $scores = {
               5_prime_title => { 17 => 1.23e-08,
                                  50 => 7.23e-04,
                                   ... },
               3_prime_title => { 300 => 2.93e-10,
                                  327 => 6.03e-05,
                                   ... }
             }

 Comment   : This class is a wrapper class for the blast command
             line interface.

=cut

sub score_hit
{

    croak "Illegal parameter list has odd number of values" if @_ % 2;

    my %params = (
	-padding => 50,
	-min_score => 0,
	@_
    );
    
    #
    # Check the required parameters
    # 
    for my $required ( qw { -db -hit -cons_seq_5 -cons_seq_3 } ) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $params{$required};
    }

    # Variables required for the Boundary Finder
    my ($seq_5_prime, $seq_3_prime, $start, $stop, $cons_seq);
    my $BM  = BlastManager->new(-db => $params{-db});
    my $hit = $params{-hit};
    my %results;
    
    #
    # Check the required values in the blast hit
    #
    for my $required ( qw { sstart send sseqid } ) {
	croak "Required parameter '$required' not passed to score_seq"
            unless exists $hit->{$required};
    }

    #
    # Aquire the 5' and 3 prime sequences
    #

    # 5' Sequence
    $start       = $hit->{sstart} - $params{-padding};
    $start       = 0 if $start < 0;
    $stop        = $hit->{send}; # + $params{-padding};
    $seq_5_prime = $BM->get_sequence(-db     => $params{-db},
				     -seq_id => $hit->{sseqid},
				     -start  => $start,
				     -end    => $stop
    );
    
    # 3' Sequence
    $start       = $hit->{sstart}; # - $params{-padding};
    $start       = 0 if $start < 0;
    $stop        = $hit->{send} + $params{-padding};
    $seq_3_prime = $BM->get_sequence(-db     => $params{-db},
				     -seq_id => $hit->{sseqid},
				     -start  => $start,
				     -end    => $stop
    );
    
    
    $cons_seq = $params{-cons_seq_5};
    $results{$cons_seq->title} = score_seq(-seq       => $seq_5_prime, 
					   -cons_seq  => $cons_seq, 
					   -min_score => $params{-min_score}
    );

    $cons_seq = $params{-cons_seq_3};
    $results{$cons_seq->title} = score_seq(-seq       => $seq_3_prime, 
					   -cons_seq  => $cons_seq, 
					   -min_score => $params{-min_score}
    );

    return \%results;
}

#################### footer pod doc ###############################

=head1 BUGS


=head1 AUTHOR

    Nathan Elmore
    nate@elmoren.com
    http://www.elmoren.com

=head1 COPYRIGHT

=head1 SEE ALSO

=cut

1;
